site-src/concepts/guidelines.md

```diff
@@ -38,6 +38,28 @@ prescribed resolution, the following guiding principles should be applied:
   should be given to the resource appearing first in alphabetical order
   (namespace/name). For example, foo/bar would be given precedence over foo/baz.
 
+## Gracefully Handling Future API Versions
+
+An important consideration when implementing this API is how it might change in
+the future. Similar to the Ingress API before it, this API is designed to be
+implemented by a variety of different products within the same cluster. That
+means that the API version your implementation was developed with may be
+different than the API version it is used with. At a minimum, the following
+requirements must be met to ensure future versions of the API do not break your
+implementation:
+
+* Handle fields with loosened validation without crashing
+* Handle fields that have transitioned from required to optional without
+  crashing
+* DO NOT rely on webhook or CRD validation as a security mechanism. If field
```

Comment:
I'm probably missing some important nuance, but [the discussion on this (around 48m)](https://youtu.be/EUUtwv6mFGg?t=2874) in this week's meeting sounded to me like the webhook is an important security mechanism, and that's why its presence (or equivalent) should be checked by conformance tests.

Maybe these two statements aren't contradictory, but I would have expected to see documentation added about how it's required for implementors to deploy the webhook to be conformant and an explanation of why it's a potential security issue if it's not deployed.

Reply:
Ah, I see you're just moving existing docs around. Maybe best addressed separately then?

Reply:
@markmc this is an important distinction, and we should be more clear about exactly what a webhook provides. I think it's worth strongly installing the installation of a webhook and even requiring it as part of a conformant deployment, but we need to recognize that a webhook can fail or be bypassed. 

/cc @youngnick for more thoughts on how we should approach this.

Reply:
I think that we're aiming for a defence-in-depth posture here. Kubebuilder validations and webhook validations are part of it, but should not be the whole story.

However, I do think it's reasonable to expect that implementations provide equivalent functionality to the webhook, and ensure that it's part of conformance testing. Of course it's possible to bypass it, and it's not the whole story, but enabling it does increase security for API users.

Using the words "you can't rely on this" kind of implies that there's no point to having them. I think that something like "implentors should ensure that, even if the protections are bypassed, their implementations are still as secure as possible" or something, particularly for things like filtering fields.

The validation protects against mistakes, the webhook against attackers without a way to bypass them, but nothing protects against a dedicated, motivated attacker with high cluster access.

Reply:
Updated, thanks!